### ISSUE TITLE
TPM flushing context error(0x901)

### ISSUE THREAD

**Issue by Ivanhu5866** [tpm2-software/tpm2-tools#2279]
We are testing TPM with real hardwares with the below script which basically followed the tpm2-tools changeauth.sh test. 

```
#!/bin/bash
ownerPasswd=abc123
endorsePasswd=abc123
lockPasswd=abc123
new_ownerPasswd=newpswd
new_endorsePasswd=newpswd
new_lockPasswd=newpswd

for i in 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 
do 
	echo "loop $i"
	rm -rf key.ctx key.pub key.priv primary.ctx policy.nvchange session.ctx >> /dev/null
	sudo tpm2_startup -c --tcti=device:/dev/tpmrm0
        sudo tpm2_clear --tcti=device:/dev/tpmrm0
# Test changing an NV index auth
	sudo tpm2_startauthsession -S session.ctx --tcti=device:/dev/tpmrm0
	sudo tpm2_policycommandcode -S session.ctx -L policy.nvchange TPM2_CC_NV_ChangeAuth --tcti=device:/dev/tpmrm0
	sudo tpm2_flushcontext session.ctx --tcti=device:/dev/tpmrm0
	NVIndex=0x1500015
	sudo tpm2_nvdefine $NVIndex -C o -s 32 -a "authread|authwrite" -L policy.nvchange --tcti=device:/dev/tpmrm0
	sudo tpm2_startauthsession --policy-session -S session.ctx --tcti=device:/dev/tpmrm0
	sudo tpm2_policycommandcode -S session.ctx -L policy.nvchange TPM2_CC_NV_ChangeAuth --tcti=device:/dev/tpmrm0
	sudo tpm2_changeauth -p session:session.ctx -c $NVIndex newindexauth --tcti=device:/dev/tpmrm0

done
```

And after the loop 12, it will get the error on Linux dmesg, then unable to run util flushing manually by command.
```
[ 707.066243] tpm tpm0: tpm2_save_context: failed with a TPM error 0x0901
[ 707.073171] tpm tpm0: A TPM error (459) occurred flushing context
[ 707.074876] tpm tpm0: A TPM error (459) occurred flushing context
[ 707.076659] tpm tpm0: A TPM error (459) occurred flushing context
[ 707.076665] tpm tpm0: tpm2_commit_space: error -14

tpm:warn(2.0): gap for context ID is too large
```

This error only occurred on those machines,
1.  the TPM2_PT_CONTEXT_GAP_MAX value is 0xFF, not spec recommended at least 0xFFFF
```
TPM2_PT_CONTEXT_GAP_MAX:
  raw: 0xFF
```
2. Using in-kernel RM(commands with --tcti=device:/dev/tpmrm0)

Any idea why the in-kernel RM will cause this issue, but using default(/dev/tpm0) won't ?
Does in-kernel RM need to add the management for the issue?

Version
OS: Ubuntu 20.04
Tpm2-tools: 4.1.1
tpm2-tss: 2.3.2

**Comment by williamcroberts** [comment]
Yeah I think this is a limitation of the in-kernel RM not handling GAP correctly. @AndreasFuchsSIT mentioned this in a call the other day, can you elaborate?

**Comment by AndreasFuchsSIT** [comment]
Well, the in-kernel RM (last I looked) does not correctly perform GAP handling. Even though there is a TCG spec on RM behavior here (https://trustedcomputinggroup.org/wp-content/uploads/TSS_2p0_TAB_ResourceManager_v1p0_r18_04082019_pub.pdf) the in-kernel RM does not implement it.

I suggest tpm2-abrmd as an RM instead. Or maybe you are more lucky than me convincing the in-kernel RM to implement the spec...

Also, you should start closing your sessions at the end of the loop using tpm2_flushcontext, because otherwise you will start failing after the 64th iteration when ACTIVE_SESSIONS_MAX is exhausted. Should also work as a work-around for your current GAP-Problem.

The TPM2_PT_CONTEXT_GAP_MAX value is something you have to discuss with your TPM vendor, as they are using the wrong value.

I'll close this issue, since it's third party bugs... Sorry... ;-)

**Comment by Ivanhu5866** [comment]
@AndreasFuchsSIT 
Thanks for the info.
Actually, this test script is only for reproducing the 0x0901 issue when we do the full tpm2-tools integration tests. 
Two more questions,
1. Where I can send a request to convince the in-kernel RM to implement the spec?
2. Is that possible to add the tpm2_flushcontext to the tpm2-tools integration tests?

**Comment by AndreasFuchsSIT** [comment]
1. The tpmdd_devel mailinglist and the maintainer Jarkko

2. @williamcroberts This is for you...

**Comment by PeterHuewe** [comment]
TPMdd Mailing List:
L:	linux-integrity@vger.kernel.org

**Comment by williamcroberts** [comment]
> 1. The tpmdd_devel mailinglist and the maintainer Jarkko
> 2. @williamcroberts This is for you...

What's for me?

**Comment by AndreasFuchsSIT** [comment]
@williamcroberts This is for you:
>     2. Is that possible to add the tpm2_flushcontext to the tpm2-tools integration tests?

**Comment by williamcroberts** [comment]
> @williamcroberts This is for you:
> 
> > ```
> > 2. Is that possible to add the tpm2_flushcontext to the tpm2-tools integration tests?
> > ```

Are they proposing the for loop example above to be added as a test? I'm confused as their are already a host of tpm2_flushcontext tests. I'm not opposed to adding tests if they cover some specific part of the expected behavior defined by the spec.

**Comment by Ivanhu5866** [comment]
Thanks for all your info.

@williamcroberts 
When we do the full integration tests on those TPM2_PT_CONTEXT_GAP_MAX value sets to 0xFF platforms with in-kernel RM, we will get the TPM error 0x0901 returned and then get fail for those related tests.
Though TPM vendor is fixing the TPM2_PT_CONTEXT_GAP_MAX value,  I think it might be a good idea to add the flushcontext  -s (maybe) to the cleanup function to those TPM2_PT_CONTEXT_GAP_MAX related tests.

**Comment by williamcroberts** [comment]
> Thanks for all your info.
> 
> @williamcroberts
> When we do the full integration tests on those TPM2_PT_CONTEXT_GAP_MAX value sets to 0xFF platforms with in-kernel RM, we will get the TPM error 0x0901 returned and then get fail for those related tests.
> Though TPM vendor is fixing the TPM2_PT_CONTEXT_GAP_MAX value, I think it might be a good idea to add the flushcontext -s (maybe) to the cleanup function to those TPM2_PT_CONTEXT_GAP_MAX related tests.

TL;DR
If you want to submit a patch, ill be more than happy to take it.

While you can run the test suite against hardware, it's still not advised and your mileage may vary (this seems to be one of those cases). This is why early on in the project I took and continue to take the stance that the test suite is designed to run against the simulator. Does the simulator have bugs? yes, but I can trivially upgrade those as they keep pace with the TPM spec and have a gold standard to develop too. While the goal is to support hardware TPM's, theirs so many bugs its hard to keep the workarounds in the code base and tested so regressions don't appear.

**Comment by Ivanhu5866** [comment]
Understood. Thanks for all your infomation.
